### DISCUSSION TITLE
how to use render function to return router-view-slot?

### DISCUSSION THREAD

**Discussion by hezhongfeng** [vuejs/router#2152] [Help and Questions] [answered]
I want use  render function to achieve the same effect.
```
<router-view v-slot="{ Component }">
  <component :is="Component" />
</router-view>
```

my code is 
```
<script>
import { h, resolveComponent } from 'vue';

export default {
  setup() {
    return () =>
      h(resolveComponent('router-view'), null, {
        default: ({ Component }) => h('component', { is: Component })
      });
  }
};
</script>
```

the resule is blank page

```
// this is ok, but I want use slot to use keep-alive 
return () => h(resolveComponent('router-view'));
```

thanks.

**Comment by skirtles-code** [comment, marked as answer]
Here's an example:

- [Playground](https://play.vuejs.org/#eNqFVFtv0zAU/ismPCSV2kTA4CHqpg00aSBuGggeyB6yxm2yJbblS1cU5b9zbMeuW1KQ2ibnOxd/53zH7aOubEj6IKI8ajpGuUQ9WnFcSnzFGBrQmtMOxVuF44KMAZwqifnoSTNr7t06z/lKxlKbWxBfNYHvrCAIpUrgxKZbu6OKyCR+DmnxLJpHYzpwW4oVb5i88Kf0qJ5m16NbU/FjQx7n4/uPBj+F0QvPuSB4Z9IqvC5VC+maiMBSsWRmDYQ4mJwgAM4vHObRX85GqE5iUm7jeYhpNCTUI0lzFGcxGuZjxfiGdjiezcOkGC1QfIBMlynvAQxrXRkgLHYXGkBRCw4ciWrbE0z1uFxA0qN3FCZLMJFoMGd424jmDnHvd/Zl0A/4gc8y8+KBpnb0kxv3CXeU/75phITHfAQtpWn1xnw9PyOxWzsHjLvnFlOP5iDOI35Jj7YhpJAY5WvLLp+inNhJG3oid7P1C8NKWRvpvR4rN8nc9zBODxQ+TrNST+X6NoLRgwyDvkLhKPb3yG54eJs4Xh/dJ7ixlAiJOrFB59qfxDe4bSn6SXlbPYtB8UBZMCTuWAszAQuhZf3iou9N8jAsM7AM2hCmJNouOlrh9ryIwF9E4FpmPhtIH+gywTrL/nu2HYUphJQwqCVxeJLtf9GVDDaSEjjLzL0YHaKIcqdEEe13T8NFVEvJRJ5lirDHTQqCZPuIy7P0Zfo6q2AzAjTFolvcc/okzBUoolHOIrqEoKzCW0lpKxYla04d8Vfg5Zv0DE5qm/sMqmcNqfDO1HY3ENqUArRcN5ujJvUONS3mX5hsQOuDZkuQ+umDwSRX2BNd1Xj1OIE/iJ2l/JVj6G6Lg+ZkyTdYWvf1t894B+/eCaugWoj+h/MWC9oqzdGGvVWkAtpBnGH73mjWkM13cb2TmAjXlCZqpmHijZD6T+xU63u6r9IzP8XhD8QzVHU=)

In that example, I'm importing `RouterView` directly, rather than using `resolveComponent`, but either way should work.

In general, `<component :is="foo">` becomes `h(foo)`. But in the specific case of `RouterView`, the `Component` slot prop is already a VNode, so the extra `h()` call can be skipped:

```js
h(RouterView, null, ({ Component }) => Component)
```

If you wanted to write that out in full it might be something like this:

```js
h(RouterView, null, {
  default({ Component }) {
    return Component ? h(Component) : null
  }
})
```

The code above also needs to take account of the initial value of `Component` being `undefined`.

**Comment by hezhongfeng** [comment]
thanks for your help.
